### PR TITLE
Tests and routing improvements

### PR DESCRIPTION
**Pull request by enricofer**
I implemented a mock_point_validity class that simutates the presence of an island in the middle of the track path and forces the algorithm to find a path that bypasses the simulated island.
For easy map estimation of paths I implemented a pathAsGeojson function in utils module. The second checkRoute test make use of this new function, printing on screen the geojson definition of the retrieved path.
About this I would ask you to specify what are the attributes returned with path waypoints
Furthermore I ask you check the path, past the resulting geojson in https://geojson.io/ , even trying to change the mocked grib params (starttws, starttwd, fuzziness) and the insland dimension factor in mock_point_validity.
It seems that the final waypoints approaching the end track point are not well calculated, particularly for low tws: 
1: https://gist.github.com/enricofer/ab54569c361bca7af8441ae250325ba9
2: https://gist.github.com/enricofer/ff404cd5a42efdfcb9cca06f7ac182ae
3: https://gist.github.com/enricofer/c7947725fd1fd211ba54075e57d65087
4: https://gist.github.com/enricofer/f0db356473aa864e661b590784555bd7

### CODE REVIEW

File: weatherrouting/utils.py
Comment (dakk) [inline weatherrouting/utils.py:154]: I think this should returns the gj object instead of the string, so this function is more generic and suitable for other usecases where a geojson object is required. Move the dumps in your print line.

Please comment out all print function, since they're only for debug, or use logging instead like this, so it can be turned on / off with one line of code:

```
log = logging.getLogger('test_2')

log.debug ('What you want to print for debugging')
```

File: tests/routing.py
Comment (dakk) [inline tests/routing.py:99]: Please comment out all print function, since they're only for debug, or use logging instead like this, so it can be turned on / off with one line of code:

```
log = logging.getLogger('test_2')

log.debug ('What you want to print for debugging')
```

File: weatherrouting/routers/linearbestisorouter.py
Comment (dakk) [inline weatherrouting/routers/linearbestisorouter.py:37]: This should use the minIncrease value, since routing should be customizable for that

File: weatherrouting/routers/router.py
Comment (dakk) [inline weatherrouting/routers/router.py:67]: Default set lineValidity and pointValidity to None before the if-elif

File: weatherrouting/routers/router.py
Comment (dakk) [inline weatherrouting/routers/router.py:102]: Use an elif; they could be both none

**Comment by enricofer** [comment]
Hi Davide, sorry for the noise. I should have to explain better the meaning of the PR, but Yesterday I was a bit spleepy, and furthermore I found some issues in the PR. Perhaps It would be better to close it.
Anyway, I'm proposing you two improvements:
1) The user provided pointValidity function is expected to check the validity of a point against a sea/terrain model. But it certain conditions, for in a context with little islands (<20 km), even if the waypoints lay on water the resulting route could intersect the coastline. So i'm proposing you to leave to the end user to check the validity against a point ( 2 params)  or a line (4 params)
2) I noticed that the actual algorithm, with low wind conditions have a wrong end condition: https://gist.github.com/enricofer/c7947725fd1fd211ba54075e57d65087 In this gist you can notice that the last calculated waypoint is too far from end point. This is way less visible with high wind condition: https://gist.github.com/enricofer/f0db356473aa864e661b590784555bd7.
I found that a fixed maxDistance param, used as exit condition does not take into account the impossibility to reach the end point with the actual wind conditions. So I'm proposing to check the distance from the end point against the maximum distance reachable with the actual wind condition. Apparently the new exit condition resolvs the issue: https://gist.github.com/enricofer/979dd242abb61c57a46e348d2eeebb51
For doing this I implemented the new "maxReachDistance " polar class method https://github.com/enricofer/libweatherrouting/blob/8a9d69074408c1f8be37d774fe3a847793524d76/weatherrouting/polar.py#L173 
Furthermore, in case of line validity, I also introduce a line of sight with end point check: https://github.com/enricofer/libweatherrouting/blob/8a9d69074408c1f8be37d774fe3a847793524d76/weatherrouting/routers/linearbestisorouter.py#L38 but I verified that sometimes the check fails (for example with costrainted geographic context like this:  
Screenshot (46) look to the end point on the right ) so a your deep revision is needed.

File: weatherrouting/routers/linearbestisorouter.py
Comment (enricofer) [reply weatherrouting/routers/linearbestisorouter.py:37]: see issue reported below

File: weatherrouting/routers/router.py
Comment (enricofer) [reply weatherrouting/routers/router.py:67]: the provided pointValidity function could have many issues, could be implemented or not, and have mismatching params in type, number and position. I think that It would better to implement an headache preventing check but it have to be more comprehensive.

File: weatherrouting/routers/router.py
Comment (enricofer) [reply weatherrouting/routers/router.py:102]: see above comment

**Comment by dakk** [comment]
1. I think we can actually give the opportunity to provide both point and line validity
2. Ok with that, it seems reasonable; anyway, could be better to use this condition only on final point?

**Comment by enricofer** [comment]
> 
> 
>     1. I think we can actually give the opportunity to provide both point and line validity
> 
>     2. Ok with that, it seems reasonable; anyway, could be better to use this condition only on final point?

Do you mean that we can do a double ckeck, first on maxDistance param, then on maxRangeDistance ?

File: weatherrouting/utils.py
Comment (enricofer) [reply weatherrouting/utils.py]: https://github.com/dakk/libweatherrouting/pull/2/commits/7bd70ea31b324bfec603372cb3d033120f769e92

**Comment by enricofer** [comment]
> 
> 
> > ```
> > 1. I think we can actually give the opportunity to provide both point and line validity
> > 
> > 2. Ok with that, it seems reasonable; anyway, could be better to use this condition only on final point?
> > ```
> 
> Do you mean that we can do a double ckeck, first on maxDistance param, then on maxRangeDistance ?

https://github.com/dakk/libweatherrouting/pull/2/commits/5eb13d0a1db181770f95f97b851fd22e0b42eb10
seems resolving

**Comment by dakk** [comment]
@enricofer 

> > > ```
> > > 1. I think we can actually give the opportunity to provide both point and line validity
> > > 
> > > 2. Ok with that, it seems reasonable; anyway, could be better to use this condition only on final point?
> > > ```
> > 
> > 
> > Do you mean that we can do a double ckeck, first on maxDistance param, then on maxRangeDistance ?
> 
> [5eb13d0](https://github.com/dakk/libweatherrouting/commit/5eb13d0a1db181770f95f97b851fd22e0b42eb10)
> seems resolving

Regarding ```1. I think we can actually give the opportunity to provide both point and line validity```, I was thinking something like this:

```python
def __init__ (self, polar, grib, pointValidity = None, lineValidity = None):
   ...
   self.lineValidity = lineValidity
   self.pointValidity = pointValidity
```

Then when we need to use point/line validity, we check both; the use case for this is if:
- we want to check if the point lays on ground (pointValidity)
- we want to check for instance that the line does not lays on natural reserve (lineValidity)

The condition will be (so both can be None, both can be defined):

```python
if (not self.pointValidity or self.pointValidity(a, b)) and (not self.lineValidity or self.lineValidity(a,b,c,d))
```

**Comment by enricofer** [comment]
So you are aware that the actual Routing class init method is the following:
`def __init__ (self, algorithm, polar, track, grib, pointValidity, startDatetime, startPosition = None):`
and should become something like this:
`def __init__ (self, algorithm, polar, track, grib, startDatetime, startPosition = None, pointValidity = None, lineValidity = None):`
in other words as pointValidity become keyworded it has to change position, and then parameters order has to change and you will have an api break in your application

**Comment by dakk** [comment]
> So you are aware that the actual Routing class init method is the following:
> `def __init__ (self, algorithm, polar, track, grib, pointValidity, startDatetime, startPosition = None):`
> and should become something like this:
> `def __init__ (self, algorithm, polar, track, grib, startDatetime, startPosition = None, pointValidity = None, lineValidity = None):`
> in other words as pointValidity become keyworded it has to change position, and then parameters order has to change and you will have an api break in your application

Yep, but it's not a problem; in the first stage api should be malleable to fit better user needs, I'll fix gwr after merge

**Review by dakk — APPROVED** [review]

**Comment by dakk** [comment]
The pull is ok for me now; if you agree I can merge

**Comment by enricofer** [comment]
For me the PR is closed, I ask you and @pcav to revise README.md

**Comment by pcav** [comment]
@enricofer do you mean https://github.com/dakk/libweatherrouting/blob/master/README.md ?

**Comment by enricofer** [comment]
No the one in my own repo fork

Il mar 20 apr 2021, 18:57 Paolo Cavallini ***@***.***> ha
scritto:

**Comment by enricofer** [comment]
> @enricofer do you mean https://github.com/dakk/libweatherrouting/blob/master/README.md ?

https://github.com/enricofer/libweatherrouting

**Comment by dakk** [comment]
ok waiting for pcav reply

**Comment by pcav** [comment]
Done, I created a PR (minor stuff only however).

**Comment by enricofer** [comment]
Davide, would you add some info on routing algorithm? Or some bibliographic reference?
Don't forget to update travis badge pointing to your repo.
Thanks all for the support.
For me is enough.

**Comment by dakk** [comment]
@enricofer will do; the project is based on https://web.archive.org/web/20180324153950/https://riccardoapolloni.altervista.org/ work

**Comment by pcav** [comment]
Interesting, thanks. I pointed out some theoretical approaches here: https://github.com/faunalia/qnavigate/blob/master/README.md#further-info

**Pull request merged**
